ta/pkcs11/include/pkcs11_ta.h

```diff
@@ -565,8 +565,87 @@ enum pkcs11_ta_cmd {
 	 * This command relates to the PKCS#11 API function C_GenerateRandom().
 	 */
 	PKCS11_CMD_GENERATE_RANDOM = 42,
+
+	/*
+	 * PKCS11_CMD_DIGEST_INIT - Initialize a digest computation processing
+	 *
+	 * [in]  memref[0] = [
+	 *              32bit session handle,
+	 *              (struct pkcs11_attribute_head)mechanism +
+	 *                                            mechanism params,
+	 *	 ]
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 *
+	 * These commands relate to the PKCS#11 API function C_DigestInit().
+	 */
+	PKCS11_CMD_DIGEST_INIT = 44,
+
+	/*
+	 * PKCS11_CMD_DIGEST_KEY - Update digest with a key
+	 *
+	 * [in]  memref[0] = [
+	 *              32bit session handle,
+	 *              32bit key handle,
+	 *	 ]
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 *
+	 * These commands relate to the PKCS#11 API function C_DigestKey().
+	 */
+	PKCS11_CMD_DIGEST_KEY = 45,
+
+	/*
+	 * PKCS11_CMD_DIGEST_UPDATE - Update digest with data
+	 *
+	 * [in]  memref[0] = [
+	 *              32bit session handle,
+	 *              32bit command flags,
+	 *	 ]
+	 * [out] memref[0] = 32bit return code, enum pkcs11_rc
+	 * [in]  memref[1] = input data to be processed
+	 *
+	 * This command relates to the PKCS#11 API function C_DigestUpdate().
+	 */
+	PKCS11_CMD_DIGEST_UPDATE = 46,
+
+	/*
+	 * PKCS11_CMD_DIGEST_FINAL - Finalize a digest computation processing
+	 *
+	 * [in]  memref[0] = [
+	 *              32bit session handle,
+	 *              32bit command flags,
```

Comment:
Why using command flags?  this command is expected to output  either the disgest size value or the digest with its size value.

Reply:
```
 A call to C_DigestFinal always terminates the active digest operation unless it returns CKR_BUFFER_TOO_SMALL or is a successful call (i.e., one which returns CKR_OK) to determine the length of the buffer needed to hold the message digest.
```
There is requirement to terminate the operation on bad arguments or with any other error. The bad argument case here is `C_DigestFinal(session, NULL, NULL);`. In order to terminate the operation we need to go to TA side.

Reply:
I would prefer we address this specific case in libckteec. We already did such thing if i'm right.
Let's not pollute the TA API for some weird API use, no?

Reply:
Oh! I get your point. It's true I faced this in xtest and needed to trying thing on other mechanisms processing.
hmmm, what about a specific TA command: `PKCS11_CMD_RELEASE_ACTIVE_PROCESSING` to cover all these cases.
@ruchi393, @Emantor, would you agree?
Pkcs11 TA tends to reflect cryptoki API, but I don't like these commands flags things. Do we really did them?

Reply:
PKCS11_CMD_RELEASE_ACTIVE_PROCESSING would be a clean way to do this.
I just checked - even Sign/Encrypt/Decrypt and Verify have the same requirement in the specification which we have missed. It would be cleaner to call "PKCS11_CMD_RELEASE_ACTIVE_PROCESSING" in case of errors in these APIs in libckteec.

Reply:
I'll try the extra command.